_posts/en/newsletters/2023-02-08-newsletter.md

```diff
@@ -54,12 +54,80 @@ Bitcoin infrastructure software.
+      (only those are considered reachable, even if other network
+      types are actually available).
+  a0link="https://bitcoincore.reviews/26847#l-22"
+
+  q1="How is an address received over the p2p network treated depending
+      on whether the address's network is reachable vs. non-reachable --
+      do we store it (add it to `AddrMan`) and/or forward it to peers?"
+  a1="If its network is reachable, we relay the address to two
+      randomly-chosen peers, else we relay it to 1 or 2 peers (whether
+      1 or 2 is randomly-chosen).
+      We only store the address if its network is reachable."
+  a1link="https://bitcoincore.reviews/26847#l-51"
+
+  q2="How can a node currently get stuck with only unreachable addresses
+      in `AddrMan`, finding no outbound peers? How does this PR fix it?"
+  a2="If the `-onlynet` configuration changes. For example, suppose the
+      node has always been run with `-onlynet=ipv4` so its AddMan has
+      no ipv6 addresses. Then the node is restarted with `-onlynet=ipv6`.
+      The fixed seeds have some ipv6 addresses, but without the PR, the node
+      won't use them since the `AddrMan` isn't _completely_ empty (it has some
+      ipv4 addresses from before). With the PR, the startup code will add
+      some ipv6 fixed seeds, since `AddrMan` doesn't contain any addresses
+      _of that_ (now reachable) _network type_ .
+  a2link="https://bitcoincore.reviews/26847#l-98"
+
+  q3="When an address we'd like to add to `AddrMan` collides with an existing
+      address, what happens? Is the existing address always dropped in favor
+      of the new address?"
+  a3="No, generally the existing address is retained (not the new one),
+      unless the existing address is deemed "terrible"
+      (see `AddrInfo::IsTerrible()`).
+  a3link="https://bitcoincore.reviews/26847#l-100"
+
+  q4="Why would it be beneficial to have an outbound connection to each
+      reachable network at all times?"
+  a4="A selfish reason is that it's harder to eclipse-attack the node,
+      since the attacker would need to run nodes on multiple networks.
+      A non-selfish reason is that it helps keep the overall network
+      together, avoiding chain splits caused by network partitions.
+      If half the nodes, including miners, ran with `-onlynet=x` and
+      the other half, including miners, ran `-onlynet=y`, then two
+      chains could emerge."
+  a4link="https://bitcoincore.reviews/26847#l-114"
+
+  q5="Why is the current logic in `ThreadOpenConnections()`, even with
+      the PR, insufficient to guarantee that the node has an outbound
+      connection to each reachable network at all times?"
+  a5="Nothing in the PR _guarantees_ any particular distribution of
+      peers among the reachable networks. For example, if we had 10k
+      clearnet addresses and only 1k I2P addresses in `AddrMan`, it's
+      likely that all our peers will be clearnet (ipv4 or ipv6)."
```

Comment:
would use a much smaller number of I2P addresses in the example (maybe 50 or 100). Otherwise, with 10 outbound peers, the chance to draw at least one peer at the given ratio would actually be significant.

Reply:
Yes, I actually thought of this while writing this part. But I was trying to stay true to the IRC chat log. But I agree, it's probably better to give an accurate answer here, fixed.

Reply:
If helpful, here are some real-world numbers from my node at the moment for the alternative networks. The clearnet numbers are relatively low because recently the node has been running onlynet=onion/i2p/cjdns and so the number of non-IsTerrible clearnet peers gradually drops.

<img width="477" alt="Screenshot 2023-02-03 at 14 16 33" src="https://user-images.githubusercontent.com/2415484/216721334-2ef8c5c7-2edf-44aa-832d-27617e6b7a7f.png">